modules/library/src/main/scala/zio/elasticsearch/DocumentResult.scala

```diff
@@ -0,0 +1,226 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.elasticsearch
+
+import zio.json.ast.Json
+import zio.json.ast.Json.Obj
+import zio.prelude.{Validation, ZValidation}
+import zio.schema.Schema
+import zio.{IO, Task, ZIO}
+
+sealed trait AggregationsResult {
+  def aggregation(name: String): Task[Option[ElasticAggregationResponse]]
+
+  def aggregations: Task[Map[String, ElasticAggregationResponse]]
+}
+
+sealed trait DocumentResult[F[_]] {
+  def documentAs[A: Schema]: Task[F[A]]
+}
+
+sealed trait DocumentWithAggregationsResult {
+  def aggregation(name: String): Task[Option[ElasticAggregationResponse]]
+
+  def aggregations: Task[Map[String, ElasticAggregationResponse]]
+
+  def documentAs[A: Schema]: Task[List[A]]
+}
+
+// todo maybe extract aggregations like `Item`, but there can be some confusion about naming
+final class AggregationResult private[elasticsearch] (private val aggregationsJson: Option[Json])
+    extends AggregationsResult {
+  def aggregation(name: String): Task[Option[ElasticAggregationResponse]] = ZIO.fromEither {
+    aggregationsJson match {
+      case Some(aggregations) =>
+        Obj.decoder.decodeJson(aggregations.toString) match {
+          case Right(res) =>
+            Validation
+              .validateAll(res.fields.toList.find { case (field, _) => field == name }.map { case (field, data) =>
+                ZValidation.fromEither(
+                  field match {
+                    case str if str.contains("terms") =>
+                      TermsAggregationResponse.decoder
+                        .decodeJson(data.toString)
+                        .map(field.split("#")(1) -> _)
+                    case _ =>
+                      Left(
+                        DecodingException(
+                          s"Could not parse all aggregations successfully."
+                        )
+                      )
+                  }
+                )
+              })
+              .toEitherWith { errors =>
+                DecodingException(
+                  s"Could not parse aggregation successfully: ${errors.mkString(",")})"
+                )
+              }
+              .map(_.toMap.get(name))
+          case Left(e) =>
+            Left(DecodingException(s"Could not parse aggregation successfully: $e)"))
+        }
+      case None =>
+        Right(None)
+    }
+  }
+
+  def aggregations: Task[Map[String, ElasticAggregationResponse]] = ZIO.fromEither {
+    aggregationsJson match {
+      case Some(aggregations) =>
+        Obj.decoder.decodeJson(aggregations.toString) match {
+          case Right(res) =>
+            Validation
```

Comment:
Can we extract this to `private def`? We have the same/similar code for `aggregation`.